src/components/CodeMirrorEditor.jsx

```diff
@@ -0,0 +1,90 @@
+import CodeMirror from 'codemirror';
+import isNil from 'lodash-es/isNil';
+import LRU from 'lru-cache';
+import PropTypes from 'prop-types';
+import React, {useEffect, useRef} from 'react';
+
+import 'codemirror/mode/htmlmixed/htmlmixed';
+import 'codemirror/mode/css/css';
+import 'codemirror/mode/javascript/javascript';
+
+const CODEMIRROR_MODES_MAP = {
+  html: 'htmlmixed',
+  css: 'css',
+  javascript: 'javascript',
+};
+
+export default function CodeMirrorEditor({
+  language,
+  projectKey,
+  source,
+  onInput,
+}) {
+  const containerRef = useRef();
+  const editorRef = useRef();
+  const docsCacheRef = useRef(new LRU(3));
+  const mode = CODEMIRROR_MODES_MAP[language];
+
+  useEffect(() => {
+    const {current: container} = containerRef;
+
+    const editor = editorRef.current = CodeMirror(
+      container,
```

Comment:
In this case, I don't think you need to unwrap the reference at all. `containerRef.current` would work fine here

Reply:
Something about refs makes me _always_ want to unwrap them, basically so I don't have to look at the ref itself any more than necessary. Also, in some cases it actually does matter (e.g. in effect cleanup functions) so I think it's simpler to just have a global rule of "always unwrap ref values into their own variable" and then you don't have to worry about it.